src/smtml/expr.ml

```diff
@@ -138,22 +140,65 @@ let rec ty (hte : t) : Ty.t =
 
 let rec is_symbolic (v : t) : bool =
   match view v with
-  | Val _ -> false
+  | Val _ | Loc _ -> false
   | Symbol _ -> true
-  | Loc _ -> false
   | Ptr { offset; _ } -> is_symbolic offset
-  | List vs -> List.exists is_symbolic vs
-  | App (_, vs) -> List.exists is_symbolic vs
-  | Unop (_, _, v) -> is_symbolic v
-  | Binop (_, _, v1, v2) -> is_symbolic v1 || is_symbolic v2
+  | Unop (_, _, v) | Cvtop (_, _, v) | Extract (v, _, _) | Binder (_, _, v) ->
+    is_symbolic v
+  | Binop (_, _, v1, v2) | Relop (_, _, v1, v2) | Concat (v1, v2) ->
+    is_symbolic v1 || is_symbolic v2
   | Triop (_, _, v1, v2, v3) ->
     is_symbolic v1 || is_symbolic v2 || is_symbolic v3
-  | Cvtop (_, _, v) -> is_symbolic v
-  | Relop (_, _, v1, v2) -> is_symbolic v1 || is_symbolic v2
-  | Naryop (_, _, vs) -> List.exists is_symbolic vs
-  | Extract (e, _, _) -> is_symbolic e
-  | Concat (e1, e2) -> is_symbolic e1 || is_symbolic e2
-  | Binder (_, _, e) -> is_symbolic e
+  | List vs | App (_, vs) | Naryop (_, _, vs) -> List.exists is_symbolic vs
+
+let inline_symbol_values map e =
+  let rec aux e =
+    match view e with
+    | Val _ | Loc _ -> e
+    | Symbol symbol -> Option.value ~default:e (Symbol.Map.find_opt symbol map)
+    | Ptr e ->
+      let offset = aux e.offset in
+      make @@ Ptr { e with offset }
+    | List vs ->
+      let vs = List.map aux vs in
+      make @@ List vs
+    | App (x, vs) ->
+      let vs = List.map aux vs in
+      make @@ App (x, vs)
+    | Unop (ty, op, v) ->
+      let v = aux v in
+      make @@ Unop (ty, op, v)
+    | Binop (ty, op, v1, v2) ->
+      let v1 = aux v1 in
+      let v2 = aux v2 in
+      make @@ Binop (ty, op, v1, v2)
+    | Triop (ty, op, v1, v2, v3) ->
+      let v1 = aux v1 in
+      let v2 = aux v2 in
+      let v3 = aux v3 in
+      make @@ Triop (ty, op, v1, v2, v3)
+    | Cvtop (ty, op, v) ->
+      let v = aux v in
+      make @@ Cvtop (ty, op, v)
+    | Relop (ty, op, v1, v2) ->
+      let v1 = aux v1 in
+      let v2 = aux v2 in
+      make @@ Relop (ty, op, v1, v2)
```

Comment:
Say you have `x = 42`, won't this build the expression `42 = 42`? Maybe we should use smart constructors? That way we get `True`, or is this ok for what you're doing?